File: frontend/src/stores/projectMember.ts
import { computed, ref } from 'vue'
import { defineStore } from 'pinia'
import type { ProjectMember, ProjectMemberWithUser, Permission, Role } from '@/types/auth'
import { projectMemberService } from '@/api/services/projectMember'

export const useProjectMemberStore = defineStore('projectMember', () => {
  const currentProjectMember = ref<ProjectMember | null>(null)
  const projectMembers = ref<Map<string, ProjectMemberWithUser[]>>(new Map())
  const isLoading = ref(false)
  const error = ref<string | null>(null)

  const currentRole = computed(() => currentProjectMember.value?.role)
  const currentPermissions = computed(() => currentProjectMember.value?.permissions ?? [])

  const isAdmin = computed(() => currentRole.value === 'admin')

  const loadProjectMembers = async (projectId: string): Promise<void> => {
    isLoading.value = true
    error.value = null

    try {
      const members = await projectMemberService.getProjectMembers(projectId)
      projectMembers.value.set(projectId, members)
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to load project members'
      throw err
    } finally {
      isLoading.value = false
    }
  }

  const loadMyProjectRole = async (projectId: string): Promise<void> => {
    isLoading.value = true
    error.value = null

    try {
      const member = await projectMemberService.getMyProjectRole(projectId)
      currentProjectMember.value = member
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to load project role'
      currentProjectMember.value = null
      throw err
    } finally {
      isLoading.value = false
    }
  }

  const addProjectMember = async (
    projectId: string,
    email: string,
    role: Role,
    permissions: Permission[]
  ): Promise<void> => {
    isLoading.value = true
    error.value = null

    try {
      await projectMemberService.addProjectMember(projectId, { email, role, permissions })
      await loadProjectMembers(projectId)
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to add project member'
      throw err
    } finally {
      isLoading.value = false
    }
  }

  const updateProjectMember = async (
    projectId: string,
    memberId: string,
    role: Role,
    permissions: Permission[]
  ): Promise<void> => {
    isLoading.value = true
    error.value = null

    try {
      await projectMemberService.updateProjectMember(projectId, memberId, { role, permissions })
      await loadProjectMembers(projectId)
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to update project member'
      throw err
    } finally {
      isLoading.value = false
    }
  }

  const removeProjectMember = async (projectId: string, memberId: string): Promise<void> => {
    isLoading.value = true
    error.value = null

    try {
      await projectMemberService.removeProjectMember(projectId, memberId)
      await loadProjectMembers(projectId)
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to remove project member'
      throw err
    } finally {
      isLoading.value = false
    }
  }

  const hasPermission = (permission: Permission | Permission[]): boolean => {
    if (!currentProjectMember.value) return false
    if (isAdmin.value) return true

    const perms = Array.isArray(permission) ? permission : [permission]
    return perms.every((p) => currentPermissions.value.includes(p))
  }

  const hasAnyPermission = (permissions: Permission[]): boolean => {
    if (!currentProjectMember.value) return false
    if (isAdmin.value) return true

    return permissions.some((p) => currentPermissions.value.includes(p))
  }

  const hasRole = (roles: Role | Role[]): boolean => {
    if (!currentProjectMember.value) return false

    const roleArray = Array.isArray(roles) ? roles : [roles]
    return roleArray.includes(currentProjectMember.value.role)
  }

  const clearCurrentProject = (): void => {
    currentProjectMember.value = null
    error.value = null
  }

  const getProjectMembers = (projectId: string): ProjectMemberWithUser[] => {
    return projectMembers.value.get(projectId) ?? []
  }

  return {
    currentProjectMember,
    projectMembers,
    isLoading,
    error,
    currentRole,
    currentPermissions,
    isAdmin,
    loadProjectMembers,
    loadMyProjectRole,
    addProjectMember,
    updateProjectMember,
    removeProjectMember,
    hasPermission,
    hasAnyPermission,
    hasRole,
    clearCurrentProject,
    getProjectMembers,
  }
})

File: backend/internal/domain/entities/project_member.go
package entities

import (
	"time"

	"github.com/google/uuid"
)

type ProjectMember struct {
	ID          uuid.UUID   `json:"id" db:"id"`
	ProjectID   uuid.UUID   `json:"project_id" db:"project_id"`
	UserID      uuid.UUID   `json:"user_id" db:"user_id"`
	Role        Role        `json:"role" db:"role"`
	Permissions Permissions `json:"permissions" db:"permissions"`
	CreatedAt   time.Time   `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time   `json:"updated_at" db:"updated_at"`
}

type ProjectMemberResponse struct {
	ID          uuid.UUID   `json:"id"`
	ProjectID   uuid.UUID   `json:"project_id"`
	UserID      uuid.UUID   `json:"user_id"`
	Role        Role        `json:"role"`
	Permissions Permissions `json:"permissions"`
	CreatedAt   time.Time   `json:"created_at"`
	UpdatedAt   time.Time   `json:"updated_at"`
}

func (pm *ProjectMember) ToResponse() ProjectMemberResponse {
	permissions := pm.Permissions
	if permissions == nil {
		permissions = Permissions{}
	}
	return ProjectMemberResponse{
		ID:          pm.ID,
		ProjectID:   pm.ProjectID,
		UserID:      pm.UserID,
		Role:        pm.Role,
		Permissions: permissions,
		CreatedAt:   pm.CreatedAt,
		UpdatedAt:   pm.UpdatedAt,
	}
}

func (pm *ProjectMember) HasPermission(perm Permission) bool {

	if pm.Role == RoleAdmin {
		return true
	}

	for _, p := range pm.Permissions {
		if p == perm {
			return true
		}
	}
	return false
}

func (pm *ProjectMember) HasAnyPermission(perms ...Permission) bool {
	for _, perm := range perms {
		if pm.HasPermission(perm) {
			return true
		}
	}
	return false
}

func (pm *ProjectMember) HasAllPermissions(perms ...Permission) bool {
	for _, perm := range perms {
		if !pm.HasPermission(perm) {
			return false
		}
	}
	return true
}

func (pm *ProjectMember) HasRole(roles ...Role) bool {
	for _, role := range roles {
		if pm.Role == role {
			return true
		}
	}
	return false
}

func (pm *ProjectMember) IsAdmin() bool {
	return pm.Role == RoleAdmin
}
